src/vl-tooltip.js

```diff
@@ -0,0 +1,624 @@
+
+        // supplement the field with fieldDef, timeFormat and numberFormat
+        var supplementedField = supplementField(field, fieldDef, timeFormat, numberFormat);
+
+        supplementedFields.push(supplementedField);
+      });
+    }
+    // if showAllFields is false, only supplement existing fields in options.fields
+    else {
+      if (options.fields && options.fields.length > 0) {
+        options.fields.forEach(function(field) {
+          // get the fieldDef in vlSpec that matches the field
+          var fieldDef = getFieldDef(vl.spec.fieldDefs(vlSpec), field);
+
+          // supplement the field with fieldDef, timeFormat and numberFormat
+          var supplementedField = supplementField(field, fieldDef, timeFormat, numberFormat);
+
+          supplementedFields.push(supplementedField);
+        })
+      }
+    }
+
+    // TODO(zening): supplement binned fields
+
+    options.fields = supplementedFields;
+
+    return options;
+  }
+
+  /**
+  * Find a field in fields that matches a fieldDef
+  * If the fieldDef is aggregated, find a field that matches the field name and
+  * the aggregation of the fieldDef.
+  * If the fieldDef is not aggregated, find a field that matches the field name.
+  * @return the matching field, or undefined if no match was found
+  */
+  function getField(fields, fieldDef) {
+    if (!fieldDef || !fields || fields.length <= 0) return;
+
+    var matchedField = undefined;
+
+    // if aggregate, match field name and aggregate operation
+    if (fieldDef.aggregate) {
+      // try find the perfect match: field name equals, aggregate operation equals
+      fields.forEach(function(field) {
+        if (!matchedField && field.field === fieldDef.field && field.aggregate === fieldDef.aggregate) {
+          matchedField = field;
+        }
+      });
+
+      // try find the second-best match: field name equals, field.aggregate = undefined
+      if (!matchedField) {
+        fields.forEach(function(field) {
+          if (!matchedField && field.field === fieldDef.field && field.aggregate === undefined) {
+            matchedField = field;
+          }
+        });
+      }
+    }
+    // if not aggregate, just match field name
+    else {
+      fields.forEach(function(field) {
+        if (!matchedField && field.field === fieldDef.field) {
+          matchedField = field;
+        }
+      });
+    }
+
+    return matchedField;
+  }
+
+  /**
+  * Find a fieldDef that matches field
+  * A matching fieldDef should have the same field name as field.
+  * If the matching fieldDef is aggregated, the aggregation should not contradict
+  * with that of the field.
+  * @return the matching fieldDef, or undefined if no match was found
+  */
+  function getFieldDef(fieldDefs, field) {
+    if (!field || !field.field || !fieldDefs) return;
+
+    var matchedFieldDef = undefined;
+
+    // field name should match, aggregation should not disagree
+    fieldDefs.forEach(function(fieldDef) {
```

Comment:
Do you always need to run through every field?  If not, use normal `for`-loop and simply return;